alarm/alarm.py

```diff
@@ -1,58 +1,160 @@
-            ("Enter minutes till alarm (hit enter for a "
-             "standard pomodoro (25 min), 'q' for exit) "))
-        seconds = POMODORO_TIMING
 
-        if minutes.lower() == 'q':
-            print('Bye')
-            break
+def play_alarm_file(alarm_file: str, timeout: int = ALARM_DURATION_IN_SECONDS) -> None:
+    """Play alarm file for a number of seconds, got the idea for timeout
+       from here: https://stackoverflow.com/a/52995334
+    """
+    proc = multiprocessing.Process(target=playsound, args=(alarm_file,))
+    proc.start()
+    proc.join(timeout=timeout)
+    proc.terminate()
 
-        elif minutes:
-            try:
-                minutes = int(minutes)
-            except ValueError:
-                print('Minutes needs to be an integer')
-                continue
-            else:
-                seconds = minutes * 60
 
-        try:
-            countdown(seconds)
-        except KeyboardInterrupt:
-            print('Interruption, starting over ...')
+def parse_args(args):
+    """Passing in args makes this easier to test:
+       https://stackoverflow.com/a/18161115
+    """
+    parser = argparse.ArgumentParser(
+        description="Play an alarm after N minutes",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    duration_group = parser.add_mutually_exclusive_group(required=True)
+    duration_group.add_argument(
+        "-s", "--seconds", type=int, help="Number of seconds before playing alarm"
+    )
+    duration_group.add_argument(
+        "-m", "--minutes", type=int, help="Number of minutes before playing alarm"
+    )
+
+    run_mode_group = parser.add_mutually_exclusive_group()
+    run_mode_group.add_argument(
+        "-b",
+        "--background",
+        action="store_true",
+        default=False,
+        help="Run timer in the background",
+    )
+    run_mode_group.add_argument(
+        "-d",
+        "--display_timer",
+        action="store_true",
+        default=False,
+        help="Show timer in console",
+    )
+
+    alarm_file_group = parser.add_mutually_exclusive_group()
+    alarm_file_group.add_argument(
+        "-l", "--song_library", help="Take a random song from a song library directory"
+    )
+    alarm_file_group.add_argument(
+        "-f", "--file", help="File path to song to play as alarm"
+    )
+    alarm_file_group.add_argument(
+        "-v", "--version", action="version", version=f"%(prog)s {version}"
+    )
+    return parser.parse_args(args)
+
+
+def _get_file(args) -> str:
+    """Get alarm song file from library, passed in file or environment"""
+    if args.song_library:
+        music_files = list(Path(args.song_library).rglob("*.mp[34]"))
+        if not music_files:
+            raise AlarmFileException(f"No music files found in {args.song_library}")
+        return str(random.choice(music_files))
+    elif args.file:
+        return args.file
+    else:
+        return os.environ["ALARM_MUSIC_FILE"]
```

Comment:
this will fail if it is not set. Also, `args.file` could be an illegal choice if it points to a non existing file. Maybe we want to test all these and throw an AlarmFileException?

Reply:
nice, done